pkg/virtual/syncer/builder/build.go

```diff
@@ -82,23 +83,38 @@ func BuildVirtualWorkspace(
 			withoutRootPathPrefix := strings.TrimPrefix(urlPath, rootPathPrefix)
 
 			// Incoming requests to this virtual workspace will look like:
-			//  /services/syncer/root:org:ws/<sync-target-name>/clusters/*/api/v1/configmaps
+			//  /services/syncer/root:org:ws/<sync-target-name>/<sync-target-uid>/clusters/*/api/v1/configmaps
 			//                  └───────────────────────────┐
 			// Where the withoutRootPathPrefix starts here: ┘
-			parts := strings.SplitN(withoutRootPathPrefix, "/", 3)
-			if len(parts) < 2 || parts[0] == "" || parts[1] == "" {
+			parts := strings.SplitN(withoutRootPathPrefix, "/", 4)
+			if len(parts) < 3 || parts[0] == "" || parts[1] == "" || parts[2] == "" {
 				return
 			}
+			workspace := parts[0]
 			workloadCusterName := parts[1]
+			syncTargetUID := parts[2]
 			apiDomainKey := dynamiccontext.APIDomainKey(clusters.ToClusterAwareKey(logicalcluster.New(parts[0]), workloadCusterName))
 
+			// In order to avoid conflicts with reusing deleted synctarget names, let's make sure that the synctarget name and synctarget UID match, if not,
+			// that likely means that a syncer is running with a stale synctarget that got deleted.
+			syncTarget, exists, err := wildcardKcpInformers.Workload().V1alpha1().SyncTargets().Informer().GetIndexer().GetByKey(clusters.ToClusterAwareKey(logicalcluster.New(workspace), workloadCusterName))
+			if !exists || err != nil {
+				runtime.HandleError(fmt.Errorf("failed to get synctarget %s|%s: %w", workspace, workloadCusterName, err))
+				return
+			}
+			syncTargetObj := syncTarget.(*workloadv1alpha1.SyncTarget)
+			if string(syncTargetObj.UID) != syncTargetUID {
+				runtime.HandleError(fmt.Errorf("sync target UID mismatch: %s != %s", syncTargetObj.UID, syncTargetUID))
+				return
+			}
+
 			realPath := "/"
-			if len(parts) > 2 {
-				realPath += parts[2]
+			if len(parts) > 3 {
+				realPath += parts[3]
 			}
 
-			//  /services/syncer/root:org:ws/<sync-target-name>/clusters/*/api/v1/configmaps
-			//                  ┌───────────────────────────────────┘
+			//  /services/syncer/root:org:ws/<sync-target-uid>/<sync-target-name>/clusters/*/api/v1/configmaps
```

Comment:
the order surprises me. Didn't notice earlier. Can we switch it around? ws/name/uid.

Reply:
ic, it's just a typo here in the comment. Can be follow-up to fix.

Reply:
ok